crates/scroll/hardforks/src/hardfork.rs

```diff
@@ -0,0 +1,272 @@
+
+use alloy_chains::{Chain, NamedChain};
+use reth_ethereum_forks::{hardfork, ChainHardforks, EthereumHardfork, ForkCondition, Hardfork};
+#[cfg(feature = "serde")]
+use serde::{Deserialize, Serialize};
+
+hardfork!(
+    /// The name of the Scroll hardfork
+    ///
+    /// When building a list of hardforks for a chain, it's still expected to mix with
+    /// [`EthereumHardfork`].
+    ScrollHardfork {
+        /// EIP150 HF block (nil = no fork)
+        EIP150Block,
+        /// EIP155 HF block
+        EIP155Block,
+        /// EIP158 HF block
+        EIP158Block,
+        /// Archimedes switch block (nil = no fork, 0 = already on archimedes)
+        Archimedes,
+        /// Bernoulli: <https://scroll.io/blog/blobs-are-here-scrolls-bernoulli-upgrade>.
+        Bernoulli,
+        /// Curie: <https://scroll.io/blog/compressing-the-gas-scrolls-curie-upgrade>.
+        Curie,
+        /// Darwin: <https://scroll.io/blog/proof-recursion-scrolls-darwin-upgrade>.
+        Darwin,
+        /// DarwinV2 <https://x.com/Scroll_ZKP/status/1830565514755584269>.
+        DarwinV2,
+    }
+);
+
+impl ScrollHardfork {
+    /// Retrieves the activation block for the specified hardfork on the given chain.
+    pub fn activation_block<H: Hardfork>(self, fork: H, chain: Chain) -> Option<u64> {
+        // will replace scroll_sepolia after alloy-chains release new tag
+        if chain == Chain::from_named(NamedChain::ScrollSepolia) {
+            return Self::scroll_sepolia_activation_block(fork);
+        }
+        if chain == Chain::from_named(NamedChain::Scroll) {
+            return Self::scroll_mainnet_activation_block(fork);
+        }
+
+        None
+    }
+
+    /// Retrieves the activation timestamp for the specified hardfork on the given chain.
+    pub fn activation_timestamp<H: Hardfork>(self, fork: H, chain: Chain) -> Option<u64> {
+        if chain == Chain::from_named(NamedChain::ScrollSepolia) {
+            return Self::scroll_sepolia_activation_timestamp(fork);
+        }
+        if chain == Chain::from_named(NamedChain::Scroll) {
+            return Self::scroll_mainnet_activation_timestamp(fork);
+        }
+
+        None
+    }
+
+    /// Retrieves the activation block for the specified hardfork on the Scroll Sepolia testnet.
+    pub fn scroll_sepolia_activation_block<H: Hardfork>(fork: H) -> Option<u64> {
+        match_hardfork(
+            fork,
+            |fork| match fork {
+                EthereumHardfork::Homestead |
+                EthereumHardfork::Byzantium |
+                EthereumHardfork::Constantinople |
+                EthereumHardfork::Petersburg |
+                EthereumHardfork::Istanbul |
+                EthereumHardfork::Berlin |
+                EthereumHardfork::London |
+                EthereumHardfork::Shanghai => Some(0),
+                _ => None,
+            },
+            |fork| match fork {
+                Self::EIP150Block | Self::EIP155Block | Self::EIP158Block | Self::Archimedes => {
```

Comment:
Should be moved to `Tangerine` and `SpuriousDragon`

Reply:
fixed [5aea8e8](https://github.com/scroll-tech/reth/pull/56/commits/5aea8e8f9a41db16b531d92a1995fc75e70fa828)